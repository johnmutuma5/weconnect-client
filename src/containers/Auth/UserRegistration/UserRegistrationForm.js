import React from 'react';
import { withRouter } from 'react-router-dom';
import Form, {formProcessComplete, formProcessFailed, formInput} from '../../../components/UI/utils/Form';
import {createAccount} from '../../../store/resources/auth';
import Modal from '../../../components/UI/Modal/Modal';
import Prompt from '../Prompt/Prompt';

// import './UserRegistrationForm.css';

export class UserRegistrationForm extends React.Component {

    constructor(props) {
        super();
        this.state = {
            values: {
                first_name: '',
                last_name: '',
                username: '',
                password: '',
                mobile: '',
                email: '',
                gender: ''
            }
        }
    }

    onSubmit(e) {
        e.preventDefault()
        console.log(this.state.values)
        createAccount(this.state.values)
            .then((response) => {
                const msg = response.msg;
                const next = '/businesses';
                this.setState(formProcessComplete(msg), this.redirect(next));
            })
            .catch((error) => {
                const msg = error.msg;
                this.setState(formProcessFailed(msg));
            });
    }

    redirect(path) {
        // returns a timed-out callback function for setState
        return () => {
            setTimeout(()=>this.props.handleReadyToLogin() ,1500);
        }
    }

    render() {
        const modalTitle = 'Register a user account';
        let form = <Form
                        formContext = { this }
                        loadElements = { this.loadElements.bind(this) }
                        onSubmit = { this.onSubmit.bind(this) } />
        return (
            <Modal title={ modalTitle } isVisible={this.props.isVisible}>
                <Prompt
                    ask={'Already have an account?'}
                    handleClicked={ this.props.handleReadyToLogin} >
                    { 'Login' }</Prompt>
                { form }
            </Modal>
        );
    }

    loadElements() {
        return [
            {
                elementType: 'fieldset',
                key: 'usrreg1',
                children: [
                    {
                        elementType: 'legend',
                        children: 'Basic and Contact Data'
                    },
                    formInput('text', 'first_name', this.state.values.first_name, 'First Name'),
                    formInput('text', 'last_name', this.state.values.last_name, 'Last Name'),
                    // formInput('text', 'gender', this.state.values.gender, 'Gender'),
                    {
                        elementType: 'select',
                        attributes: {
                            name: 'gender'
                        },
                        children: [
                            {
                                elementType: 'option',
                                attributes: {
                                    value: 'Male'
                                },
                                children: 'Male'
                            },
                            {
                                elementType: 'option',
                                attributes: {
                                    value: 'Female'
                                },
                                children: 'Female'
                            }
                        ]
                    },
                    formInput('email', 'email', this.state.values.email, 'Email'),
                    formInput('text', 'mobile', this.state.values.mobile, '254700000000')
                ]
            },
            {
                elementType: 'fieldset',
                key: 'usrreg2',
                children: [
                    {
                        elementType: 'legend',
                        children: 'Login Details'
                    },
                    formInput('text', 'username', this.state.values.username, 'Username'),
                    formInput('password', 'password', this.state.values.password, 'Password')
                ]
            },
            {
                elementType: 'fieldset',
                key: 'usrreg3',
                children: [
                    formInput('submit', 'submit', 'Submit', undefined)
                ]
            }
        ]
    }
}

export default withRouter(UserRegistrationForm);
