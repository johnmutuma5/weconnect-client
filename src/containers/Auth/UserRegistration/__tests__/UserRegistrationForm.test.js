import { UserRegistrationForm } from '../UserRegistrationForm';

jest.mock('../../../../store/resources/auth')

describe('<UserRegistrationForm />', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<UserRegistrationForm />);
        expect(wrapper).toMatchSnapshot();
    });

    describe('UserRegistrationForm with simulated data', () => {
        let wrapper, username;
        beforeEach(() => {
            username = 'john_doe';
            wrapper = mount(<UserRegistrationForm isVisible />);
            wrapper.find('input[name="username"]')
                .simulate('change', {target: {name: 'username', value: username}});
            wrapper.find('input[name="password"]')
                .simulate('change', {target: {name: 'password', value: 'apassword'}});
        })

        it ('collects data from input fields into state', () => {
            expect.assertions(3);
            expect(wrapper).toMatchSnapshot();
            expect(wrapper.instance().state.values.username).toBe('john_doe');
            expect(wrapper.instance().state.values.password).toBe('apassword');
        });

        it('submits data and receives response asynchronously', (done) => {
            expect.assertions(1);
            wrapper.find('form')
                .simulate('submit');
            // wait for the next async tick
            setTimeout(() => {
                const successMessage = wrapper.instance().state.success_message;
                expect(successMessage).toEqual(`SUCCESS: user ${username} created!`);
                done();
            });
        });
    })
});
