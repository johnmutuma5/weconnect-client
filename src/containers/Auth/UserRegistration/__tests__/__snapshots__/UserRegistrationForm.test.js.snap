// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserRegistrationForm /> UserRegistrationForm with simulated data collects data from input fields into state 1`] = `
<div
  className="Modal visible"
>
  <p>
    Register a user account
  </p>
  <span />
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "margin": "15px 0",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      Already have an account?
    </div>
    <div
      className="Button dark"
    >
       
      Login
    </div>
  </div>
  <article
    className="form_container"
  >
    <form
      onSubmit={[Function]}
    >
      <fieldset
        key="usrreg1"
      >
        <legend
          key="Basic and Contact Data"
        >
          Basic and Contact Data
        </legend>
        <input
          key="first_name"
          name="first_name"
          onChange={[Function]}
          placeholder="First Name"
          type="text"
          value=""
        />
        <input
          key="last_name"
          name="last_name"
          onChange={[Function]}
          placeholder="Last Name"
          type="text"
          value=""
        />
        <input
          key="gender"
          name="gender"
          onChange={[Function]}
          placeholder="Gender"
          type="text"
          value=""
        />
        <input
          key="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          key="mobile"
          name="mobile"
          onChange={[Function]}
          placeholder="Mobile No."
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        key="usrreg2"
      >
        <legend
          key="Login Details"
        >
          Login Details
        </legend>
        <input
          key="username"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value="john_doe"
        />
        <input
          key="password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value="apassword"
        />
      </fieldset>
      <fieldset
        key="usrreg3"
      >
        <input
          key="submit"
          name="submit"
          onChange={[Function]}
          type="submit"
          value="Submit"
        />
      </fieldset>
    </form>
  </article>
</div>
`;

exports[`<UserRegistrationForm /> renders correctly 1`] = `
<modal
  title="Register a user account"
>
  <prompt
    ask="Already have an account?"
  >
    Login
  </prompt>
  <Form
    formContext={
      UserRegistrationForm {
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "values": Object {
            "email": "",
            "first_name": "",
            "gender": "",
            "last_name": "",
            "mobile": "",
            "password": "",
            "username": "",
          },
        },
        "updater": Updater {
          "_callbacks": Array [],
          "_renderer": ReactShallowRenderer {
            "_context": Object {},
            "_element": <UserRegistrationForm />,
            "_forcedUpdate": false,
            "_instance": [Circular],
            "_newState": null,
            "_rendered": <modal
              isVisible={undefined}
              title="Register a user account"
      >
              <prompt
                      ask="Already have an account?"
                      handleClicked={undefined}
              >
                      Login
              </prompt>
              <Form
                      formContext={[Circular]}
                      loadElements={[Function]}
                      onSubmit={[Function]}
              />
      </modal>,
            "_rendering": false,
            "_updater": [Circular],
          },
        },
      }
    }
    loadElements={[Function]}
    onSubmit={[Function]}
  />
</modal>
`;
